sammba/modality_processors/perfusion_fair.py

```diff
@@ -0,0 +1,369 @@
+        
+    outtype : {'simple', 'complicated'}, optional
+        Changes return type. If 'simple', return a list of parameter, rCBF and
+        absolute CBF values. If 'complicated', return the two classes of
+        scipy.optimize.OptimizeResult plus rCBF and CBF values.
+    
+    Returns
+    -------
+    If outtype='simple':
+        List of 8 floats- selective inversion bias, M0 and T1, non-selective
+        inversion bias, M0 and T1, rCBF and CBF. Failed fits produce zeroes.
+    If outtype='complicated':
+        List of 4- two scipy.optimize.OptimizeResult classes (selective fit
+        then non-selective) plus floats for rCBF then CBF.
+    """
+    
+    # there are no standard errors as these are not in the current output of
+    # scipy.optimize.least_squares. the lmfit library can produce them, but
+    # during testing they were usually incalculable, and lmfit was twice as
+    # slow (or maybe I was not using it properly?) it is unlikely that the
+    # errors are of much use any way so will stay with
+    # scipy.optimize.least_squares.
+    
+    if outtype not in ['simple', 'complicated']: return 'unrecognized outtype'
+    
+    s0_sel = np.array(s0)[np.array(picker_sel)]
+    s0_nonsel = np.array(s0)[np.array(picker_nonsel)]
+
+    r_sel = fair_t1_fit(s0_sel, ti, t1_guess)
+    r_nonsel = fair_t1_fit(s0_nonsel, ti, t1_guess)
+    
+    if r_sel.success and r_nonsel.success:
+        
+        t1_sel = r_sel.x[2]
+        t1_nonsel = r_nonsel.x[2]
+        
+        rCBF = 100 * (t1_nonsel - t1_sel) / t1_nonsel
+        CBF = multiplier * lambda_blood * (
+              (t1_nonsel / t1_blood) * ((1 / t1_sel) - (1 / t1_nonsel)))
+     
+        if outtype == 'simple':
+            r = [x for x in r_sel.x] + [x for x in r_nonsel.x] + [rCBF, CBF]
+            return [x if x is not None else 0 for x in r]
+        else: return [r_sel, r_nonsel, rCBF, CBF]
+    
+    else:
+        
+        if outtype == 'simple': return np.repeat(0, 8)
+        else:
+            return ['r_sel fit ' + r_sel.success,
+                    'r_nonsel fit ' + r_nonsel.success,
+                    'rCBF nan', 'CBF nan']
+
+
+def perf_fair_fitter_mp(all_s0, t1_blood, lambda_blood, ti, multiplier,
+                        t1_guess, picker_sel, picker_nonsel, ncpu):
+    """
+    Wrapper to execute perf_fair_fitter (outtype='simple') in parallel on
+    multiple signal vectors, tracking execution with a progress bar.
+    
+    Parameters
+    ----------
+    all_s0 : 2D numpy array of int or float
+        The aquired signals. The first dimension is the number of s0 vectors,
+        the second dimension the length of all of them. So for a real image,
+        usually len(D1) >> len(D2).
+        
+    t1_blood : int or float
+        T1 of blood in ms at the acquisition field strength. Empirical
+        constant; an example value is 2800 at 11.7T (Duong ref).
+        
+    lambda_blood : int or float

```

Comment:
may be can be an optional parameter with 0.9 as default value

Reply:
Yep